src/librustc_middle/ty/mod.rs

```diff
@@ -1034,7 +1053,57 @@ impl<'tcx> Eq for Predicate<'tcx> {}
 impl<'tcx> Predicate<'tcx> {
     #[inline(always)]
     pub fn kind(self) -> &'tcx PredicateKind<'tcx> {
-        self.kind
+        &self.inner.kind
+    }
+
+    /// Skips `PredicateKind::ForAll`.
+    pub fn ignore_qualifiers(self) -> Binder<Predicate<'tcx>> {
+        match self.kind() {
+            &PredicateKind::ForAll(binder) => binder,
+            ty::PredicateKind::Projection(..)
+            | ty::PredicateKind::Trait(..)
+            | ty::PredicateKind::Subtype(..)
+            | ty::PredicateKind::WellFormed(..)
+            | ty::PredicateKind::ObjectSafe(..)
+            | ty::PredicateKind::ClosureKind(..)
+            | ty::PredicateKind::TypeOutlives(..)
+            | ty::PredicateKind::ConstEvaluatable(..)
+            | ty::PredicateKind::ConstEquate(..)
+            | ty::PredicateKind::RegionOutlives(..) => {
+                // We can't use `Binder::dummy` here, as predicates can
+                // contain unbound variables in rare cases, for example when
+                // dealing with opaque types.
+                Binder::bind(self)
```

Comment:
You'll have to update the DeBruijn indices here, so that any escaping bound variables stay escaping:
```rust
if self.has_escaping_bound_vars() {
    Binder::bind(fold::shift_vars(tcx, &self, 1))
} else {
    Binder::dummy()
}
```